src/model.rs

```diff
@@ -206,62 +223,218 @@ impl Analysis<Mdl> for TensorAnalysis {
                 let t_a = x(a).meta;
 
                 unsafe {
-                    let relu = g.relu(t_a, true);
-                    //let r_cost = (*(*relu).op.ptr).runtime;
-                    //println!("Cost of relu is {}", r_cost);
+                    let res = g.relu(t_a, true);
+                    Self::Data {
+                        dtype: DataKind::Tnsr,
+                        val: 0,
+                        name: String::new(),
+                        meta: res,
+                    }
+                }
+            }
+
+            Mdl::Tanh(a) => {
+                assert!(x(a).dtype == DataKind::Tnsr);
+                let t_a = x(a).meta;
+
+                unsafe {
+                    let res = g.tanh(t_a, true);
                     Self::Data {
                         dtype: DataKind::Tnsr,
                         val: 0,
-                        meta: relu,
+                        name: String::new(),
+                        meta: res,
                     }
                 }
             }
 
-            Mdl::Input([name, dim1, dim2, dim3, dim4]) => {
+            Mdl::Sigmoid(a) => {
+                assert!(x(a).dtype == DataKind::Tnsr);
+                let t_a = x(a).meta;
+
+                unsafe {
+                    let res = g.sigmoid(t_a, true);
+                    Self::Data {
+                        dtype: DataKind::Tnsr,
+                        val: 0,
+                        name: String::new(),
+                        meta: res,
+                    }
+                }
+            }
+
+            Mdl::Input([name]) => {
+                // Check types
                 assert!(x(name).dtype == DataKind::Name);
-                assert!(x(dim1).dtype == DataKind::Scalar);
-                assert!(x(dim2).dtype == DataKind::Scalar);
-                assert!(x(dim3).dtype == DataKind::Scalar);
-                assert!(x(dim4).dtype == DataKind::Scalar);
 
+                // Get shape
+                let mut split = x(name).name.split("@");
+                let name_vec: Vec<&str> = split.collect();
+                assert!(name_vec.len() == 2);
+                let mut split_dims = name_vec[1].split("_");
+                let dim_str_vec: Vec<&str> = split_dims.collect();
+
+                // Create tensorhandle and get metadata
                 unsafe {
-                    // NOTE all this just to pass ownership
-                    // to C++, not sure if necessary
-                    let mut dims = vec![x(dim1).val, x(dim2).val, x(dim3).val, x(dim4).val];
+                    let mut dims: Vec<i32> = dim_str_vec
```

Comment:
You can write
```
let mut dims: Vec<i32> = name_vec[1].split("_").map(|x| x.parse::<i32>().unwrap()).collect();
```
This avoids a `mut` and is more efficient as you do not need to allocate a `Vec<&str>`.

Reply:
Updated